modules/core/include/opencv2/core/utils/filesystem.private.hpp

```diff
@@ -0,0 +1,64 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_UTILS_FILESYSTEM_PRIVATE_HPP
+#define OPENCV_UTILS_FILESYSTEM_PRIVATE_HPP
+
+// TODO Move to CMake?
+#ifndef OPENCV_HAVE_FILESYSTEM_SUPPORT
+#  if defined(__EMSCRIPTEN__) || defined(__native_client__)
+     /* no support */
+#  elif defined __ANDROID__ || defined __linux__ || defined _WIN32 || \
+        defined __FreeBSD__ || defined __bsdi__
+#      define OPENCV_HAVE_FILESYSTEM_SUPPORT 1
+#  elif defined(__APPLE__)
+#    include <TargetConditionals.h>
+#    if (defined(TARGET_OS_OSX) && TARGET_OS_OSX) || (!defined(TARGET_OS_OSX) && !TARGET_OS_IPHONE)
+#      define OPENCV_HAVE_FILESYSTEM_SUPPORT 1 // OSX only
+#    endif
+#  else
+     /* unknown */
+#  endif
+#  ifndef OPENCV_HAVE_FILESYSTEM_SUPPORT
+#    define OPENCV_HAVE_FILESYSTEM_SUPPORT 0
+#  endif
+#endif
+
+#if OPENCV_HAVE_FILESYSTEM_SUPPORT
+namespace cv { namespace utils { namespace fs {
+
+/**
+ * File-based lock object.
+ *
+ * Provides interprocess synchronization mechanism.
+ * Platform dependent.
+ *
+ * Supports multiple readers / single writer access pattern (RW / readers–writer / shared-exclusive lock).
+ *
+ * File must exist.
+ * File can't be re-used (for example, I/O operations via std::fstream is not safe)
+ */
+class CV_EXPORTS FileLock {
+public:
+    explicit FileLock(const char* fname);
+    ~FileLock();
+
+    void lock(); //< acquire exclusive (writer) lock
+    void unlock(); //< release exclusive (writer) lock
+
+    void lock_sharable(); //< acquire sharable (reader) lock
```

Comment:
We are trying to use mixed case names for methods everywhere in OpenCV. Can we do it here as well?

Reply:
This is not OpenCV specific name. Also this is not a part of external public API (not available on some platforms).

Renamed to `lock_shared()` (typo is here) which is used in:
- [C++14](http://en.cppreference.com/w/cpp/thread/shared_lock/lock)
- [Boost](http://www.boost.org/doc/libs/1_51_0/doc/html/thread/synchronization.html#thread.synchronization.mutex_concepts.shared_lockable.lock_shared)